playground/serve.go

```diff
@@ -16,27 +16,237 @@
+	// The name chosen includes the timestamp of the last known
+	// components list (or 0 for the first attempt), which means
+	// that only one update task can be created for each version
+	// known.
+	t.Name = fmt.Sprintf("amp-components-list-last-known-%d", timestamp)
+	minBackoff, _ := time.ParseDuration("10s")
+	t.RetryOptions = &taskqueue.RetryOptions{
+		RetryLimit: 5,
+		MinBackoff: minBackoff,
+	}
+	log.Infof(ctx, "Adding to taskqueue: %s", t.Name)
+	taskqueue.Add(ctx, t, "")
+}
+
+func componentsTask(w http.ResponseWriter, r *http.Request) {
+	ctx := appengine.NewContext(r)
+	_, err := updateComponents(ctx)
+	if err != nil {
+		http.Error(w, "Server error", http.StatusInternalServerError)
+		return
+	}
+	w.Write([]byte("Updated components"))
+}
+
+func updateComponents(ctx context.Context) (bool, error) {
+	log.Infof(ctx, "Fetching components list from GitHub")
+	list, err := fetchComponents(ctx)
+	if err != nil {
+		return false, err
+	}
+	addComponentsToStores(ctx, list)
+	return true, nil
+}
+
+func getComponentsFromDataStore(ctx context.Context) (*AmpComponentsList, error) {
+	log.Infof(ctx, "Retrieving components from datastore")
+	var list AmpComponentsList
+	key := datastore.NewKey(ctx, "AmpComponentsList", "",
+		COMPONENTS_DATASTORE_KEY_ID, nil)
+	err := datastore.Get(ctx, key, &list)
+	if err != nil {
+		log.Infof(ctx, "Error retrieving components from datastore")
+		return nil, err
+	}
+	return &list, err
+}
+
+func getComponentsFromMemCache(ctx context.Context) (*AmpComponentsList, error) {
+	log.Infof(ctx, "Retrieving components from memcache")
+	var item AmpComponentsList
+	_, err := memcache.Gob.Get(ctx, COMPONENTS_MEMCACHE_KEY, &item)
+	if err == memcache.ErrCacheMiss {
+		log.Infof(ctx, "Components not in memcache, retrieving from datastore")
+		dsItem, err := getComponentsFromDataStore(ctx)
+		// Add to datastore
+		if err == nil {
+			log.Infof(ctx, "Setting datastore components value to memcache")
+			memcacheItem := &memcache.Item{
+				Key:    COMPONENTS_MEMCACHE_KEY,
+				Object: dsItem,
+			}
+			memcache.Gob.Set(ctx, memcacheItem)
+		}
+	} else if err != nil {
+		return nil, err
+	}
+	return &item, err
+}
+
+func addComponentsToStores(ctx context.Context, components []byte) {
+	var list AmpComponentsList
+	list.Components = components
+	list.Timestamp = int(time.Now().Unix())
+	list.Id = COMPONENTS_DATASTORE_KEY_ID
+	key := datastore.NewKey(ctx, "AmpComponentsList", "",
+		COMPONENTS_DATASTORE_KEY_ID, nil)
+	log.Infof(ctx, "Adding components to datastore")
+	_, err := datastore.Put(ctx, key, list)
+	if err != nil {
+		log.Infof(ctx, "Adding components to memcache")
+		item := &memcache.Item{
+			Key:    COMPONENTS_MEMCACHE_KEY,
+			Object: list,
+		}
+		memcache.Gob.Set(ctx, item)
+	}
+}
+
+func extractComponents(g *GitHubApiResponse) ([]byte, error) {
+	vers := make(map[string]string)
+	tree := g.Tree
+	for _, blob := range tree {
+		groups := componentRegex.FindStringSubmatch(blob.Path)
```

Comment:
suggestion: extract the parsing into a separate method. Then you can avoid the deep nesting:

```
func parseAndAddComponent(blob, componentsWithVersion) {
                groups := componentRegex.FindStringSubmatch(blob.Path)
 		if len(groups) != 3 {
                     return
                }
 	       component := groups[1]
 		if strings.HasSuffix(component, "-impl") {
                   return
                }
 		ver := groups[2]
 		elem, ok := vers[component]
 		if !ok || ver > elem {
 		  componentsWithVersion[component] = ver
 		}
 	}
}
```

Reply:
Done